cpp/src/fil/infer.cu

```diff
@@ -91,10 +95,40 @@ __device__ __forceinline__ vec<1, output_type> infer_one_tree(tree_type tree,
   return out;
 }
 
+template <int NITEMS>
+struct best_margin_label {
+  vec<NITEMS, float> margin;
+  vec<NITEMS, int> label;
+};
+
+template <int NITEMS>
+struct ArgMax {
+  __device__ best_margin_label<NITEMS> operator()(
```

Comment:
This function looks more complicated than necessary. Could you introduce `struct best_margin_label` with just 1 margin and 1 label, overload `operator>()` for it, and implement this function in terms of it?

The structure I'm thinking of could look like the following.

```
struct best_margin_label {
  float margin;
  int label;
  best_margin_label(float margin_, int label_) : margin(margin_), label(label_) {}
  friend bool operator>(best_margin_label a, best_margin_label b) {
    return a.margin > b.margin || a.margin == b.margin && a.label < b.label;
  }
};
```